File: sdk/typescript/src/exec.ts
/**
 * Process executor — spawns `claude-code-rs` and reads JSONL from stdout.
 *
 * Modeled after `codex/sdk/typescript/src/exec.ts`.
 */

import { spawn } from "node:child_process";
import path from "node:path";
import readline from "node:readline";
import { execSync } from "node:child_process";

import type { PermissionMode } from "./sessionOptions.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ClaudeCodeExecArgs = {
  input: string;
  apiKey?: string;
  model?: string;
  workingDirectory?: string;
  permissionMode?: PermissionMode;
  maxTurns?: number;
  maxBudget?: number;
  systemPrompt?: string;
  appendSystemPrompt?: string;
  verbose?: boolean;
  continueSession?: string;
  signal?: AbortSignal;
};

const INTERNAL_ORIGINATOR_ENV = "CC_RUST_INTERNAL_ORIGINATOR";
const SDK_ORIGINATOR = "claude_code_rs_sdk_ts";

// ---------------------------------------------------------------------------
// Executor
// ---------------------------------------------------------------------------

export class ClaudeCodeExec {
  private executablePath: string;
  private envOverride?: Record<string, string>;

  constructor(
    executablePath?: string | null,
    env?: Record<string, string>,
  ) {
    this.executablePath = executablePath || findClaudeCodeRsPath();
    this.envOverride = env;
  }

  async *run(args: ClaudeCodeExecArgs): AsyncGenerator<string> {
    const commandArgs: string[] = ["--output-format", "json", "-p"];

    if (args.model) {
      commandArgs.push("--model", args.model);
    }
    if (args.workingDirectory) {
      commandArgs.push("--cwd", args.workingDirectory);
    }
    if (args.permissionMode) {
      commandArgs.push("--permission-mode", args.permissionMode);
    }
    if (args.maxTurns !== undefined) {
      commandArgs.push("--max-turns", String(args.maxTurns));
    }
    if (args.maxBudget !== undefined) {
      commandArgs.push("--max-budget", String(args.maxBudget));
    }
    if (args.systemPrompt) {
      commandArgs.push("--system-prompt", args.systemPrompt);
    }
    if (args.appendSystemPrompt) {
      commandArgs.push("--append-system-prompt", args.appendSystemPrompt);
    }
    if (args.verbose) {
      commandArgs.push("--verbose");
    }
    if (args.continueSession) {
      commandArgs.push("--continue", args.continueSession);
    }

    // Build environment
    const env: Record<string, string> = {};
    if (this.envOverride) {
      Object.assign(env, this.envOverride);
    } else {
      for (const [key, value] of Object.entries(process.env)) {
        if (value !== undefined) {
          env[key] = value;
        }
      }
    }
    if (!env[INTERNAL_ORIGINATOR_ENV]) {
      env[INTERNAL_ORIGINATOR_ENV] = SDK_ORIGINATOR;
    }
    if (args.apiKey) {
      env.ANTHROPIC_API_KEY = args.apiKey;
    }

    const child = spawn(this.executablePath, commandArgs, {
      env,
      signal: args.signal,
    });

    let spawnError: unknown | null = null;
    child.once("error", (err) => (spawnError = err));

    if (!child.stdin) {
      child.kill();
      throw new Error("Child process has no stdin");
    }
    child.stdin.write(args.input);
    child.stdin.end();

    if (!child.stdout) {
      child.kill();
      throw new Error("Child process has no stdout");
    }

    const stderrChunks: Buffer[] = [];
    if (child.stderr) {
      child.stderr.on("data", (data: Buffer) => {
        stderrChunks.push(data);
      });
    }

    const exitPromise = new Promise<{
      code: number | null;
      signal: NodeJS.Signals | null;
    }>((resolve) => {
      child.once("exit", (code, signal) => {
        resolve({ code, signal });
      });
    });

    const rl = readline.createInterface({
      input: child.stdout,
      crlfDelay: Infinity,
    });

    try {
      for await (const line of rl) {
        yield line as string;
      }

      if (spawnError) throw spawnError;
      const { code, signal } = await exitPromise;
      if (code !== 0 || signal) {
        const stderrBuffer = Buffer.concat(stderrChunks);
        const detail = signal ? `signal ${signal}` : `code ${code ?? 1}`;
        throw new Error(
          `claude-code-rs exited with ${detail}: ${stderrBuffer.toString("utf8")}`,
        );
      }
    } finally {
      rl.close();
      child.removeAllListeners();
      try {
        if (!child.killed) child.kill();
      } catch {
        // ignore
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Binary resolution
// ---------------------------------------------------------------------------

function findClaudeCodeRsPath(): string {
  // 1. Explicit env var
  const envPath = process.env["CLAUDE_CODE_RS_PATH"];
  if (envPath) return envPath;

  // 2. Check PATH
  try {
    const cmd =
      process.platform === "win32"
        ? "where claude-code-rs"
        : "which claude-code-rs";
    const result = execSync(cmd, { encoding: "utf8" }).trim();
    if (result) return result.split("\n")[0]!.trim();
  } catch {
    // not on PATH
  }

  // 3. Relative to this package (cargo build output)
  const binaryName =
    process.platform === "win32" ? "claude-code-rs.exe" : "claude-code-rs";
  const candidates = [
    path.resolve(__dirname, "..", "..", "..", "target", "release", binaryName),
    path.resolve(__dirname, "..", "..", "..", "target", "debug", binaryName),
  ];

  for (const candidate of candidates) {
    try {
      const fs = require("node:fs");
      if (fs.existsSync(candidate)) return candidate;
    } catch {
      // continue
    }
  }

  throw new Error(
    "Unable to locate claude-code-rs binary. " +
      "Set CLAUDE_CODE_RS_PATH environment variable or ensure it is on PATH.",
  );
}

File: sdk/python/src/claude_code_rs/exec.py
"""Process executor — spawns ``claude-code-rs`` and reads JSONL from stdout.

Mirrors TypeScript SDK's exec.ts.
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
import threading
from collections import deque
from typing import Iterator

from .config import ClientOptions, SessionOptions
from .errors import BinaryNotFoundError, ParseError, ProcessError

INTERNAL_ORIGINATOR_ENV = "CC_RUST_INTERNAL_ORIGINATOR"
SDK_ORIGINATOR = "cc_rust_sdk_python"

# ---------------------------------------------------------------------------
# Binary resolution
# ---------------------------------------------------------------------------


def find_claude_code_rs_path() -> str:
    """Locate the ``claude-code-rs`` binary.

    Resolution order (mirrors TypeScript SDK exec.ts):
    1. ``CLAUDE_CODE_RS_PATH`` environment variable
    2. System PATH via ``shutil.which``
    3. Relative cargo build output (release then debug)
    """
    searched: list[str] = []

    # 1. Explicit env var
    env_path = os.environ.get("CLAUDE_CODE_RS_PATH")
    if env_path:
        return env_path

    # 2. System PATH
    which_result = shutil.which("claude-code-rs")
    if which_result:
        return which_result
    searched.append("PATH (claude-code-rs)")

    # 3. Relative to this package (cargo build output)
    binary_name = (
        "claude-code-rs.exe" if sys.platform == "win32" else "claude-code-rs"
    )
    # sdk/python/src/claude_code_rs/ → ../../.. → sdk/ → .. → project root
    package_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.normpath(
        os.path.join(package_dir, "..", "..", "..", "..")
    )
    for profile in ("release", "debug"):
        candidate = os.path.join(project_root, "target", profile, binary_name)
        if os.path.isfile(candidate):
            return candidate
        searched.append(candidate)

    raise BinaryNotFoundError(searched)


# ---------------------------------------------------------------------------
# Executor
# ---------------------------------------------------------------------------


class ClaudeCodeExec:
    """Manages spawning the ``claude-code-rs`` subprocess and reading JSONL."""

    def __init__(
        self,
        executable_path: str | None = None,
        env: dict[str, str] | None = None,
    ) -> None:
        self._executable_path = executable_path or find_claude_code_rs_path()
        self._env_override = env

    def run(
        self,
        *,
        input: str,
        api_key: str | None = None,
        model: str | None = None,
        working_directory: str | None = None,
        permission_mode: str | None = None,
        max_turns: int | None = None,
        max_budget: float | None = None,
        system_prompt: str | None = None,
        append_system_prompt: str | None = None,
        verbose: bool = False,
        continue_session: str | None = None,
    ) -> Iterator[str]:
        """Spawn the CLI, write *input* to stdin, yield stdout lines.

        This is a generator function. The subprocess is cleaned up in the
        ``finally`` block when the generator is closed or exhausted.
        """
        args = self._build_args(
            model=model,
            working_directory=working_directory,
            permission_mode=permission_mode,
            max_turns=max_turns,
            max_budget=max_budget,
            system_prompt=system_prompt,
            append_system_prompt=append_system_prompt,
            verbose=verbose,
            continue_session=continue_session,
        )
        env = self._build_env(api_key=api_key)

        proc = subprocess.Popen(
            args,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            env=env,
        )

        # Collect stderr in a background daemon thread to avoid deadlock
        stderr_lines: deque[str] = deque(maxlen=400)
        stderr_thread = threading.Thread(
            target=_drain_stderr, args=(proc, stderr_lines), daemon=True
        )
        stderr_thread.start()

        try:
            # Write prompt and close stdin
            assert proc.stdin is not None
            proc.stdin.write(input.encode("utf-8"))
            proc.stdin.close()

            # Yield stdout lines
            assert proc.stdout is not None
            for raw_line in proc.stdout:
                line = raw_line.decode("utf-8", errors="replace").rstrip("\n")
                if line:
                    yield line

            # Wait for process to finish
            exit_code = proc.wait()
            stderr_thread.join(timeout=1.0)

            if exit_code != 0:
                raise ProcessError(exit_code, "\n".join(stderr_lines))

        finally:
            # Guarantee cleanup
            if proc.poll() is None:
                proc.terminate()
                try:
                    proc.wait(timeout=2.0)
                except subprocess.TimeoutExpired:
                    proc.kill()
                    proc.wait()
            stderr_thread.join(timeout=0.5)

    # -------------------------------------------------------------------
    # Internal helpers
    # -------------------------------------------------------------------

    def _build_args(
        self,
        *,
        model: str | None,
        working_directory: str | None,
        permission_mode: str | None,
        max_turns: int | None,
        max_budget: float | None,
        system_prompt: str | None,
        append_system_prompt: str | None,
        verbose: bool,
        continue_session: str | None,
    ) -> list[str]:
        args = [self._executable_path, "--output-format", "json", "-p"]

        if model:
            args += ["--model", model]
        if working_directory:
            args += ["--cwd", working_directory]
        if permission_mode:
            args += ["--permission-mode", permission_mode]
        if max_turns is not None:
            args += ["--max-turns", str(max_turns)]
        if max_budget is not None:
            args += ["--max-budget", str(max_budget)]
        if system_prompt:
            args += ["--system-prompt", system_prompt]
        if append_system_prompt:
            args += ["--append-system-prompt", append_system_prompt]
        if verbose:
            args.append("--verbose")
        if continue_session:
            args += ["--continue", continue_session]

        return args

    def _build_env(self, *, api_key: str | None) -> dict[str, str]:
        if self._env_override is not None:
            env = dict(self._env_override)
        else:
            env = dict(os.environ)

        if INTERNAL_ORIGINATOR_ENV not in env:
            env[INTERNAL_ORIGINATOR_ENV] = SDK_ORIGINATOR
        if api_key:
            env["ANTHROPIC_API_KEY"] = api_key

        return env


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _drain_stderr(
    proc: subprocess.Popen[bytes], buf: deque[str]
) -> None:
    """Background thread: read stderr lines into *buf*."""
    assert proc.stderr is not None
    for raw_line in proc.stderr:
        buf.append(raw_line.decode("utf-8", errors="replace").rstrip("\n"))
